metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,200 @@
+package io.iohk.metronome.checkpointing.service
+
+import cats.Applicative
+import cats.data.{NonEmptyList, NonEmptyVector, OptionT}
+import cats.effect.concurrent.Ref
+import cats.effect.{Concurrent, Resource, Sync}
+import cats.implicits._
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.checkpointing.models.{
+  Block,
+  CheckpointCertificate,
+  Ledger
+}
+import io.iohk.metronome.checkpointing.service.CheckpointingService.LedgerTree
+import io.iohk.metronome.checkpointing.service.storage.LedgerStorage
+import io.iohk.metronome.crypto.ECPublicKey
+import io.iohk.metronome.hotstuff.consensus.basic.{Phase, QuorumCertificate}
+import io.iohk.metronome.hotstuff.service.ApplicationService
+import io.iohk.metronome.hotstuff.service.storage.{
+  BlockStorage,
+  ViewStateStorage
+}
+import io.iohk.metronome.storage.KVStoreRunner
+
+class CheckpointingService[F[_]: Sync: Applicative, N](
+    ledgerTree: Ref[F, LedgerTree],
+    lastExecutedHeader: Ref[F, Block.Header],
+    ledgerStorage: LedgerStorage[N],
+    blockStorage: BlockStorage[N, CheckpointingAgreement]
+)(implicit storeRunner: KVStoreRunner[F, N])
+    extends ApplicationService[F, CheckpointingAgreement] {
+
+  override def createBlock(
+      highQC: QuorumCertificate[CheckpointingAgreement]
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Option[Boolean]] = {
+    OptionT(updateLedger(block)).flatMapF { ledger =>
+      val hashValid = ledger.hash == block.header.postStateHash
+      validateTransactions(block.body, ledger).map {
+        _.map(_ && hashValid)
+      }
+    }.value
+  }
+
+  private def validateTransactions(
+      body: Block.Body,
+      ledger: Ledger
+  ): F[Option[Boolean]] = {
+    //TODO: Validate transactions PM-3131/3132
+    true.some.pure[F]
+  }
+
+  override def executeBlock(
+      block: Block,
+      commitQC: QuorumCertificate[CheckpointingAgreement],
+      commitPath: NonEmptyList[Block.Hash]
+  ): F[Boolean] = {
+    require(commitQC.phase == Phase.Commit, "Commit QC required")
+    updateLedger(block).flatMap {
+      case Some(ledger) =>
+        if (block.hash != commitQC.blockHash)
+          saveState(block.header, ledger).as(true)
+        else
+          for {
+            //TODO: PM-3110:
+            // chkpOpt <- constructCheckpoint(ledger, commitQC)
+            _ <- saveState(block.header, ledger)
+            //TODO: PM-3110:
+            // _ <- chkpOpt.map(pushCheckpoint).getOrElse(().pure[F])
+          } yield true
+
+      case None =>
+        Sync[F].raiseError(
+          //TODO: better error info with Either?
```

Comment:
Yeah, it would only be helpful if we mess up something badly. I guess we can always add some debugging code in that case. I will remove this comment

Reply:
Tracing is a nice way to add as much debug content as you need.

Reply:
Is that a suggestion for now or something to keep in mind? I did not add any tracing now, as I don't suppose this is likely to surface as an issue.

Reply:
I only mentioned it in the context of "I guess we can always add some debugging code". 

Exceptions returned by this code are traced in the `BlockExecutor`, but if you ever wanted to add actual debugging, it would be a good indication for the need to add tracing.